lib/Conversion/TorchToTosa/TorchToTosa.cpp

```diff
@@ -1760,6 +1760,166 @@ LogicalResult ConvertAtenOp<AtenConv2dOp>::matchAndRewrite(
+
+  rewriter.replaceOpWithNewOp<tosa::ReshapeOp>(
+      op, getTypeConverter()->convertType(newType), self,
+      rewriter.getI64ArrayAttr(newShape));
+
+  return success();
+}
+
+// Normalization ops perform elementwise ops of a single mean/stdev value
+// against the feature map and because input is NCHW, the rank-1 value must be
+// reshaped so it sits on the same dim as 'C'.
+static LogicalResult reshapeToNormInputDim(Operation *op,
+                                           ConversionPatternRewriter &rewriter,
+                                           TypeConverter *converter,
+                                           Type outType, const Value &toBcast,
+                                           Value &result) {
+  RankedTensorType toBcastType = toBcast.getType().dyn_cast<RankedTensorType>();
+  if (toBcastType.getRank() > 1)
+    op->emitError("Rank cannot be more than 1");
+
+  RankedTensorType outTensorType = outType.cast<RankedTensorType>();
+  SmallVector<int64_t> newShape = {toBcastType.getShape()[0]};
+  for (auto i = 2; i < outTensorType.getRank(); ++i)
+    newShape.push_back(1);
+  auto newType =
+      RankedTensorType::get(newShape, outTensorType.getElementType());
+
+  result = rewriter.create<tosa::ReshapeOp>(
+      op->getLoc(), converter->convertType(newType), toBcast,
+      rewriter.getI64ArrayAttr(newShape));
+
+  return success();
+}
+
+// This lowering is based on the TF to TOSA lowering.
+template <>
+LogicalResult ConvertAtenOp<AtenBatchNormOp>::matchAndRewrite(
+    AtenBatchNormOp op, OpAdaptor adaptor,
+    ConversionPatternRewriter &rewriter) const {
+
+  // Not a ranked tensor output
+  if (!adaptor.input().getType().dyn_cast<RankedTensorType>())
+    return op.emitError("Only ranked tensor types are supported");
+
+  auto outType = getTypeConverter()->convertType(op.getType());
+
+  // FIXME: Handle training, momentum and cudnn_enabled
+
+  // For PyTorch:
+  //   scale  = gamma = weight
+  //   offset = beta  = bias
+  // Lowering:
+  // fused batchnorm = (input-mean) * scale * rsqrt(var+epsilon)) + offset
+  //
+  // shape_0 = ones(input.rank)
+  // shape_0[input.rank-1] = input.shape[input.rank-1]
+  // shape_1 = ones(1)
+  //
+  // bmean  = reshape(mean, shape_0)
+  // bscale = reshape(scale, shape_0)
+  // boffset= reshape(offset, shape_0)
+  // beps   = reshape(epsilon, shape_1)
+  //
+  // op1 = sub(input, bmean)
+  // op2 = add(var, beps)
+  // op3 = rsqrt(op2)
+  // bvar = reshape(op3, shape_0)
+  // op4 = mul(op1, bvar)
+  // op5 = mul(op4, bscale)
+  // op6 = add(op5, boffset)
+
+  auto meanType = adaptor.running_mean().getType().dyn_cast<TensorType>();
+  auto varianceType = adaptor.running_var().getType().dyn_cast<TensorType>();
+  if (!varianceType || !meanType)
+    return op.emitError("Only ranked tensor types are supported");
+
+  Value meanVal, varianceVal, weightVal, biasVal;
+  if (meanType.getNumElements() == 0) {
```

Comment:
I think this doesn't really exist right? Maybe we could just assert the getNumElements != 0?

Reply:
Would that be the case for varianceType as well?

Reply:
Yea, I think so.